ydb/core/kqp/executer_actor/kqp_partitioned_executer.cpp

```diff
@@ -62,420 +90,805 @@ class TKqpPartitionedExecuter : public TActorBootstrapped<TKqpPartitionedExecute
+        TString paramName;
+
+        try {
+            for (const auto& paramDesc : PreparedQuery->GetParameters()) {
+                paramName = paramDesc.GetName();
+                queryData->ValidateParameter(paramDesc.GetName(), paramDesc.GetType(), request.TxAlloc->TypeEnv);
+            }
+
+            for(const auto& paramBinding: request.Transactions.front().Body->GetParamBindings()) {
+                paramName = paramBinding.GetName();
+                queryData->MaterializeParamValue(true, paramBinding);
+            }
+        } catch (const yexception& ex) {
+            RuntimeError(
+                Ydb::StatusIds::BAD_REQUEST,
+                NYql::TIssues({NYql::TIssue(TStringBuilder() << CurrentStateFuncName()
+                    << ", cannot prepare parameters for request, parameter name = " << paramName)}));
+        }
+    }
+
+    bool CheckExecutersAreFinished() const {
+        return StartedPartitions.empty();
+    }
+
+    // SchemeCache and ReadActor may have the different order of key columns,
+    // so we need to reorder partition ranges for next compares.
+    void TryReorderKeysByIds(const TVector<ui32>& keyIds) {
+        if (keyIds.empty()) {
+            return;
+        }
+
+        YQL_ENSURE(KeyColumnInfo.size() == keyIds.size());
+
+        bool isEqual = true;
+        for (size_t i = 0; i < KeyColumnInfo.size(); ++i) {
+            if (KeyColumnInfo[i].Id != keyIds[i]) {
+                isEqual = false;
+                break;
+            }
+        }
+
+        if (isEqual) {
+            return;
+        }
+
+        ReorderKeyColumnInfo(keyIds);
+        for (const auto& [partIdx, _] : StartedPartitions) {
+            ReorderPartitionRanges(partIdx);
+        }
+    }
+
+    void ReorderPartitionRanges(TPartitionIndex idx) {
+        PE_LOG_D("Reorder KeyColumnInfo and partitioning ranges by keyIds from RA");
+
+        auto& partInfo = StartedPartitions[idx];
+
+        auto& beginRow = partInfo->BeginRange;
+        if (!beginRow.Empty()) {
+            beginRow->EndKeyPrefix = ReorderRangeByKeyColumnInfo(beginRow->EndKeyPrefix);
+        }
+
+        auto& endRow = partInfo->EndRange;
+        if (!endRow.Empty()) {
+            endRow->EndKeyPrefix = ReorderRangeByKeyColumnInfo(endRow->EndKeyPrefix);
+        }
+    }
+
+    TSerializedCellVec ReorderRangeByKeyColumnInfo(const TSerializedCellVec& row) {
+        if (row.GetCells().empty()) {
+            return row;
+        }
+
+        TVector<TCell> newRow;
+        auto cells = row.GetCells();
+        for (const auto& info : KeyColumnInfo) {
+            newRow.push_back(cells[info.ParamIndex]);
+        }
+
+        TConstArrayRef<TCell> rowRef(newRow);
+        return TSerializedCellVec(rowRef);
+    }
+
+    void ReorderKeyColumnInfo(const TVector<ui32>& keyIds) {
+        TVector<TKeyColumnInfo> newInfo;
+
+        for (const auto& id : keyIds) {
+            for (const auto& info : KeyColumnInfo) {
```

Comment:
Also need to verify that you have found something.

Maybe
it = std::find(...)
YQL_ENSURE(it != ....end())
newInfo.push_back(*it)